pkg/chunk/purger/blocks_purger.go

```diff
@@ -0,0 +1,234 @@
+	if err := tManager.RemoveCancelledStateIfExists(ctx, requestID); err != nil {
+		level.Error(util_log.Logger).Log("msg", "removing cancelled tombstone state if it exists", "err", err)
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+
+	prevT, err := tManager.GetTombstoneByIDForUser(ctx, requestID)
+	if err != nil {
+		level.Error(util_log.Logger).Log("msg", "error getting delete request by id", "err", err)
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+	if prevT != nil {
+		http.Error(w, "delete request tombstone with same information already exists", http.StatusBadRequest)
+		return
+	}
+
+	curTime := time.Now().Unix() * 1000
+	t := cortex_tsdb.NewTombstone(userID, curTime, curTime, startTime, endTime, match, requestID, cortex_tsdb.StatePending)
+
+	if err = tManager.WriteTombstone(ctx, t); err != nil {
+		level.Error(util_log.Logger).Log("msg", "error adding delete request to the object store", "err", err)
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+
+	w.WriteHeader(http.StatusNoContent)
+}
+
+func (api *BlocksPurgerAPI) GetAllDeleteRequestsHandler(w http.ResponseWriter, r *http.Request) {
+	ctx := r.Context()
+	userID, err := tenant.TenantID(ctx)
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusUnauthorized)
+		return
+	}
+	tManager := cortex_tsdb.NewTombstoneManager(api.bucketClient, userID, api.cfgProvider, api.logger)
+	deleteRequests, err := tManager.GetAllTombstonesForUser(ctx)
+	if err != nil {
+		level.Error(util_log.Logger).Log("msg", "error getting delete requests from the block store", "err", err)
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+
+	if err := json.NewEncoder(w).Encode(deleteRequests); err != nil {
+		level.Error(util_log.Logger).Log("msg", "error marshalling response", "err", err)
+		http.Error(w, fmt.Sprintf("Error marshalling response: %v", err), http.StatusInternalServerError)
+	}
+
+	w.WriteHeader(http.StatusOK)
+
+}
+
+func (api *BlocksPurgerAPI) CancelDeleteRequestHandler(w http.ResponseWriter, r *http.Request) {
+	ctx := r.Context()
+	userID, err := tenant.TenantID(ctx)
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusUnauthorized)
+		return
+	}
+
+	params := r.URL.Query()
+	requestID := params.Get("request_id")
+
+	tManager := cortex_tsdb.NewTombstoneManager(api.bucketClient, userID, api.cfgProvider, api.logger)
+	deleteRequest, err := tManager.GetTombstoneByIDForUser(ctx, requestID)
+	if err != nil {
+		level.Error(util_log.Logger).Log("msg", "error getting delete request from the object store", "err", err)
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+
+	if deleteRequest == nil {
+		http.Error(w, "could not find delete request with given id", http.StatusBadRequest)
+		return
+	}
+
+	if deleteRequest.State == cortex_tsdb.StateCancelled {
+		http.Error(w, "the request has already been previously deleted", http.StatusBadRequest)
```

Comment:
In this case, shouldn't we return 202? Because it's already in cancelled state,  which is what this API should do right? 

By the way, in our error messages to the user, we gotta use "cancel" and "delete" carefully. I think Here you actually mean "the series deletion request was cancelled previously" not "the series has already been deleted" right?

Reply:
Good point, thanks.